### ISSUE TITLE
iOS 9 support

### ISSUE THREAD

**Issue by voron** [juhovh/shairplay#43]
I'm unable to use shairplay with iOS 9 beta 3. On the same time Apple Airport works fine with the same iOS 9 iPad and the same instance of shairplay works fine with iOS 8.4.1 another iPad. Here is iOS 9-related console output 
```
[root@playroom bin]# shairplay -a playroom --ao_driver=alsa --hwaddr=b8:27:eb:51:d8:a4
Initialized server socket(s)
*** WARNING *** The program 'shairplay' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=shairplay>
Accepted IPv4 client on socket 10
Local: 192.168.78.30
Remote: 192.168.78.20
Accepted IPv6 client on socket 11
Local: fe80:0000:0000:0000:0213:efff:fe40:0725
Remote: fe80:0000:0000:0000:0409:966b:bb87:92c1
Receiving on socket 10
Connection closed for socket 10
Receiving on socket 11
Got challenge: S26Dy1m46AgDAPXbUUY2Wg==
Got response: nV2dEZJabFCnF+ENxz0eFhdBUy9H/JIk4HrC5y1vuN5xh6QNyB6zzRjADosZodtabQD1Cdd7zkwEO4S9buzXceRyjsDJYRaSSmc8DioSBt4H22X0iwPP8kkbu9VR2M3lU5nZm8his6/WkdRH0jGG4zIJGDWY43FWqr+Z8OWKDUOqmx0nYdxPT2MOQvFymNXsry4o3KxKzmf65jKhZFE258QARmdpyscoF4S+gVqUZIqt0ez7ZA2vFVta5Qekdc5ehKE52aFotIoYp8Ar9E1q2eFP1Flcd2qDQ5fYjk4ShATG/WIS783LOin9GOrB9DTZKofNz556GQlRig5I1vgvsw
Handled request OPTIONS with URL *
Receiving on socket 11
connection: IN IP6 fe80::409:966b:bb87:92c1
rtpmap: 96 AppleLossless
fmtp: 96 352 0 16 40 10 14 2 255 0 0 44100
rsaaeskey: HaF9c0dc1EZIut4hK9GNqEf9Xnm2/GmXWxYx1MQ7NK8ghPkTM5F2c6OfqPjQEoMYFw30X+I+qrKWRbAI88Q2RkSe3ekZm6QnGT0VmTYTpeeDz8xOyFwOTrTE4tT40Ga1JpWJT5sWdAvzrg1NLe4Q8uUznBWlgbfjNBEg+KyhBI+GP/w2ZVzSaUWdvSna3tEQd2R6TnAHKrrTdznqJ9p01CS4kWfs2WK0MTOtEsBWzY8Qeo8g8EaVKlncBVocENXxO7HoOHVTaW6HcF0+MJG1BgjQdzvT0mB+HwE2lkH5s5ornb3GY8tmT1GgEq+tij9QpZjZQX1sSULqS3XCP6WgyA==
aesiv: T0MW8tN5nLKJscniCmALyA==
aeskeylen: 16
aesivlen: 16
Handled request ANNOUNCE with URL rtsp://fe80::213:efff:fe40:725/16485092811308287640
Receiving on socket 11
Transport: RTP/AVP/UDP;unicast;mode=record;timing_port=51410;control_port=50525
Found remote timing port: 51410
Found remote control port: 50525
Responding with RTP/AVP/UDP;unicast;mode=record;timing_port=58308;events;control_port=37767;server_port=50503
Handled request SETUP with URL rtsp://fe80::213:efff:fe40:725/16485092811308287640
Receiving on socket 11
Handled request GET_PARAMETER with URL rtsp://fe80::213:efff:fe40:725/16485092811308287640
Receiving on socket 11
Exiting UDP RAOP thread
Handled request TEARDOWN with URL rtsp://fe80::213:efff:fe40:725/16485092811308287640
Receiving on socket 11
Connection closed for socket 11
```
Do you need any additional info?

**Comment by juhovh** [comment]
Thank you, it looks like iOS 9 has started to make a GET_PARAMETER request, which is unhandled by shairplay, because so far it has not been clear what that request is used for. That is why the connection gets disconnected. Unfortunately the debug log doesn't give the whole contents of GET_PARAMETER.

It could be something very simple like the iPad just requesting volume parameter before starting the stream. Could you please attach a full dump of the handshake with for example "tcpdump -i <interface> -s 65535 -w <some-file>" and I'll have a look at it.

**Comment by voron** [comment]
Here is tcpflow http://pastebin.com/raw.php?i=e5v1kfNk and here is tcpdump of another handshake with some other packets like icmp  https://www.dropbox.com/s/a5xgerz9x3o2xzb/shairplay.tcpdump?dl=0

**Comment by juhovh** [comment]
Seems to be just requesting volume as I suspected, this might mean I have to add a new callback for requesting volume, since it's not stored anywhere. I'll let you know when I have a fix for testing.

**Comment by jvah** [comment]
@voron sorry to be late about this, but one thing came to my mind, would you mind trying to delete the "GET_PARAMETER, " text from the string on src/lib/raop.c line 205 and recompile? That might be a quick fix, but if it doesn't work I'll actually implement the GET_PARAMETER. It's just a bit annoying to implement since there can be multiple values in one request...

**Comment by juhovh** [comment]
@voron sorry to be late about this, but one thing came to my mind, would you mind trying to delete the "GET_PARAMETER, " text from the string on src/lib/raop.c line 205 and recompile? That might be a quick fix, but if it doesn't work I'll actually implement the GET_PARAMETER. It's just a bit annoying to implement since there can be multiple values in one request...

**Comment by voron** [comment]
I had recompiled shairplay without GET_PARAMETER - it's still requested and iPad disconnects from shairplay. Here is tcpflow http://pastebin.com/raw.php?i=PFxACZD9

**Comment by voron** [comment]
Strangely, there is no GET_PARAMETER with iOS9 and Airport Express http://pastebin.com/raw.php?i=X58D9C7e . This seems to be common iOS 9 related problem - airplay was re-designed, old firmware on AppleTV is broken and so on. I suspect no quick solution.

**Comment by juhovh** [comment]
Yeah, I just looked at the article in [9to5mac](http://9to5mac.com/2015/09/11/apple-ios-9-airplay-improvements-screen-mirroring/) which suggest even apps like Reflector need changes to make things work. I guess have to just wait for the smoke to settle down and then get to this, I'm worried that the work that has been put to reverse engineer screen mirroring might go to waste as well now.

**Comment by voron** [comment]
http://pastebin.com/raw.php?i=X58D9C7e tcpflow with iOS9 and Airport Express fw 7.6.3 just in case.

**Comment by livewireintegration** [comment]
I would love to get this fixed, prepared to offer a bounty also. I am not a strong developer but have isolated the issue in here:

raop_rtp.c  - Line 423

None of the below conditions appear to be met for iOS9

if (FD_ISSET(raop_rtp->csock, &rfds)) {
}
else if (FD_ISSET(raop_rtp->tsock, &rfds)) {
}
else if (FD_ISSET(raop_rtp->dsock, &rfds)) {
}

**Comment by livewireintegration** [comment]
Really keen to get this fixed. Happy to offer $'s it can motivate a faster resolution :)

**Comment by juhovh** [comment]
I finally got to update my devices to iOS 9 and I can confirm that just implementing GET_PARAMETER fixes the issue, I'll start implementing it. Screen mirroring is another issue that needs to be looked into later.

**Comment by juhovh** [comment]
Please test the latest master branch version for iOS 9 support, seems to work for me.

@voron @livewireintegration @Memphiz

**Comment by voron** [comment]
Works for me too.

**Comment by Memphiz** [comment]
Mhh this is only half of the story for kodi sadly. I had to disable the airplay announcement to make airtunes work with your changes. It seems the fact that kodi announces both roap and airplay service ios9 somehow tries to do its new connect handshake stuff which we know nothing about :(((

Good work though for the roap only users :)

**Comment by juhovh** [comment]
@Memphiz ah, that makes sense but sucks... I noticed Reflector now advertises iOS9 support at http://www.airsquirrels.com/reflector/ maybe that could give some hints too about the changes.

**Comment by juhovh** [comment]
Just out of curiosity I posted a dissected RAOP handshake of iOS 9 screen mirroring session to http://pastebin.com/raw.php?i=wLWEqLWg. Basically it seems to do: pair-verify, pair-setup, pair-verify, pair-verify. Then it proceeds with the fp-setup handshake as usual, don't know how tricky it would be to get the raop and airplay support to work concurrently... Should really get this whole process figured out so that mirroring and whatever could be made work.

**Comment by Memphiz** [comment]
Also the whole metadata stuff seems broken (no duration/progress, no artist, no album, no title, no album thumb) ...

**Comment by Memphiz** [comment]
for completness - DACP (remote control) still works with the music app ...

**Comment by juhovh** [comment]
It's very possible that since we still emulate airport device (which doesn't have a screen) then this metadata is no more sent to that class of device... Should get the Apple TV DRM working and then probably everything would work as before. But easier said than done, no one seems to have time and skills. :)

**Comment by juhovh** [comment]
By the way the pair-setup handshake seems to be related to HomeKit, there is some implementation at https://github.com/brutella/hc/blob/master/netio/endpoint/pair-verify.go that has implemented it but the documentation is quite slim. Looks like Apple has merged the AirPlay and HomeKit protocols in iOS 9 to work better with each other.

**Comment by firsttris** [comment]
there is also a working homeKit server for nodejs with iOS9 API.
https://github.com/KhaosT/HAP-NodeJS
https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/HAPServer.js

**Comment by juhovh** [comment]
It looks like HomeKit pairing is actually using SRP with TLV encoded values, but the AirPlay pairing data is quite different. Need to continue to look into this.

**Comment by mikebrady** [comment]
I must say I haven't come across that GET_PARAMETER stuff with Shairport Sync, though I wonder if it's been taken out of iOS 9?

Has anybody had any joy finding where the metadata is gone?

**Comment by kam187** [comment]
So we all know the protocol changed in ios9, so i've been digging and playing around some more.

There seems to be a fallback mode where the pair-verify etc is skipped if you use the correct feature flags:

                 'features': u'0xF79F7',
                'model' : u'AppleTV3,2',
                'srcvers': u'220.68',
                'deviceid': u':'.join(self.device_id),

Using these flags the device sends:

POST '/fp-setup' (challenge 1 as per the previous protocol)
POST '/fp-setup' (challenge 2 as per the previous protocol)

<then things change>

SETUP 'rtsp://192.168.1.3/16568656483503263028'
with content type application/x-apple-binary-plist which is a binary plist (decrypt the key here eiv/ekey):

Example Contents:
{'et': 32, 'eiv': '\xc6\x9d,L/T\xdd\xd4k_\xf1\xafi\x84\x07\x95', 'streams': [{'type': 110, 'timestampInfo': [{'name': 'SubSu'}, {'name': 'BePxT'}, {'name': 'AfPxT'}, {'name': 'BefEn'}, {'name': 'EmEnc'}], 'streamConnectionID': -7068938806120134259}], 'sessionUUID': 'E5EFB149-88AF-4D34-A7D6-7A27AD46B58C', 'osBuildVersion': '13A452', 'sourceVersion': '241.33', 'timingPort': 61021, 'ekey': 'FPLY\x01\x02\x01\x00\x00\x00\x00<\x00\x00\x00\x00\n\xddd\xf1\x1d\xa2\x8e\x0ej\xe2\xef\x14\xa6\x0bC\xdb\x00\x00\x00\x10\x96\xcfS2\xd7s\x94Q\xd7Lf\x9b-\x95 \x85\x92\xd0\xd0\xe2o\xdd\xba(\xc8g\xbck\xe5\xa7\xdc\r\x9a\xa5b\xb4', 'deviceID': 'XX:XX:XX:XX:XX:B7', 'model': 'iPhone7,2', 'name': 'iPhonw', 'macAddress': 'XX:XX:XX:XX:XX:B5'}

Note the macAddress is the Wifi address from my phone, and the device id is the next mac up from my bluetooth address, maybe it's the cell mac address who knows.

I respond with the following binary plist:
{'streams': [{'dataPort': 7100, 'type': 110}], 'eventPort': 57959, 'timingPort': 54460}

The device then tries to connect to TCP 'eventPort' and if it fails then TEARDOWN. So i open a fake port there and ignore the data.

Next the device sends GET '/info' and I respond with a binary plist again:

"{'pk': ",f<A\x0eF|\xd5\xc2~G\x00\xbf\xe5\xf7\x8d\xdeJ\x16'\\\xe4\xa6Bz|K\xf3\xf0\x1f\xc0\xba", 'name': 'FL501', 'vv': 2, 'statusFlags': 68, 'keepAliveLowPower': 1, 'keepAliveSendStatsAsBody': 1, 'pi': 'b08f5a79-db29-4384-b456-a4784d9e6055', 'sourceVersion': '220.68', 'deviceID': 'XX:XX:XX:XX:XX:XX', 'macAddress': 'XX:XX:XX:XX:XX:FE', 'model': 'AppleTV3,2', 'audioFormats': [{'audioInputFormats': 67108860, 'audioOutputFormats': 67108860, 'type': 100}, {'audioInputFormats': 67108860, 'audioOutputFormats': 67108860, 'type': 101}], 'audioLatencies': [{'audioType': 'default', 'inputLatencyMicros': False, 'outputLatencyMicros': False, 'type': 100}, {'audioType': 'default', 'inputLatencyMicros': False, 'outputLatencyMicros': False, 'type': 101}], 'features': 130367356919, 'displays': [{'height': 720, 'width': 1280, 'rotation': False, 'widthPhysical': False, 'heightPhysical': False, 'widthPixels': 1280, 'heightPixels': 720, 'refreshRate': 60, 'features': 14, 'overscanned': False, 'uuid': 'e5f7a68d-7b0f-4305-984b-974f677a150b'}]}"

The device then sends:

GET_PARAMETER 'rtsp://192.168.1.3/16568656483503263028'
'volume'

You need to respond with volume in the response NOT the header:

            response = "volume: 0.000000\x0d\x0a"
            self.send_response(200)
            self.send_header("Server", self.version_string())
            self.send_header("Content-Type", "text/parameters")
            self.send_header("Content-Length", str(len(response)))
            self.end_headers()
            self.wfile.write(response)

And finally you'll receive RECORD

That's where i'm stuck. I'm not receiving any data on any of the ports in the exchange.  I'm probably missing something obvious, but you might have a better idea.

**Comment by frdmn** [comment]
Really looking forward for iOS9 support! Keep up the good work :+1:

**Comment by notedit** [comment]
any  progress?

**Comment by mikebrady** [comment]
Shairport Sync is seeing metadata again in the RAOP protocol...

**Comment by Jack-lx** [comment]
Hi juhovh:
"pair-verify" " pair-setup",this two sessions is resolved ?can you help me to how reseponse two sessions?
Thank you very much!

**Comment by Jack-lx** [comment]
@juhovh ,
"pair-verify" " pair-setup",this two sessions is resolved ?can you help me to how reseponse two sessions?
Thank you very much!

**Comment by wenliabc2007** [comment]
@Jack-lx @juhovh @voron 
"pair-verify" " pair-setup",this two sessions is resolved ?can you help me to how reseponse two sessions?
Thank you very much!

**Comment by zhengwu1008** [comment]
"pair-verify" " pair-setup",this two sessions is resolved ?can you help me to how reseponse two sessions?
Thank you very much!

**Comment by shahasachin** [comment]
Hi Jack-Ix / wenliabc2007,
Can you please provide steps/details about how you have resolved "pair-verify" " pair-setup" sessions? 
Thank you very much
